smartsim/entity/strategies.py

```diff
@@ -90,32 +95,69 @@ def _impl(
+    # Create dictionaries for each parameter permutation
+    param_zip = [
+        dict(zip(file_params, permutation)) for permutation in file_params_permutations
+    ][:_n_permutations]
+    # Generate all possible permutations of executable arguments
+    exe_arg_params_permutations = itertools.product(*exe_arg_params.values())
+    # Create dictionaries for each executable argument permutation
+    exe_arg_zip = [
+        dict(zip(exe_arg_params, permutation))
+        for permutation in exe_arg_params_permutations
+    ][:_n_permutations]
+    # Combine parameter and executable argument dictionaries
+    combinations = itertools.product(param_zip, exe_arg_zip)
+    # Combine the parameter sets from 'param_zip' and 'exe_arg_zip' using itertools.zip_longest
+    param_set = (ParamSet(file_param, exe_arg) for file_param, exe_arg in combinations)
+    slice = itertools.islice(param_set, _n_permutations)
+    return list(slice)
 
 
 @_register("step")
 def step_values(
-    params: t.Mapping[str, t.Sequence[str]], _n_permutations: int = 0
-) -> list[dict[str, str]]:
-    steps = zip(*params.values())
-    return [dict(zip(params, step)) for step in steps]
+    params: t.Mapping[str, t.Sequence[str]],
+    exe_args: t.Mapping[str, t.Sequence[t.Sequence[str]]],
+    _n_permutations: int = 0,
+) -> list[ParamSet]:
+    # Zip the values of the 'params' dictionary
+    param_zip = zip(*params.values())
+    # Create a list of dictionaries, where each dictionary represents a combination of parameter values
+    # Limit the list to '_n_permutations' elements
+    param_zip = [dict(zip(params, step)) for step in param_zip][:_n_permutations]
+    # Zip the values of the 'exe_args' dictionary
+    exe_arg_zip = zip(*exe_args.values())
+    # Create a list of dictionaries, where each dictionary represents a combination of executable argument values
+    # Limit the list to '_n_permutations' elements
+    exe_arg_zip = [dict(zip(exe_args, step)) for step in exe_arg_zip][:_n_permutations]
+    # Combine the parameter sets from 'param_zip' and 'exe_arg_zip' using itertools.zip_longest
+    param_set = (
+        ParamSet(file_param, exe_arg)
+        for (file_param, exe_arg) in itertools.zip_longest(param_zip, exe_arg_zip)
+    )
+    # Limit the generator to '_n_permutations' elements
+    slice = itertools.islice(param_set, _n_permutations)
+    # Convert the limited generator to a list and return it
+    return list(slice)
 
 
 @_register("random")
 def random_permutations(
-    params: t.Mapping[str, t.Sequence[str]], n_permutations: int = 0
-) -> list[dict[str, str]]:
-    permutations = create_all_permutations(params, 0)
-
-    # sample from available permutations if n_permutations is specified
-    if 0 < n_permutations < len(permutations):
-        permutations = random.sample(permutations, n_permutations)
-
+    params: t.Mapping[str, t.Sequence[str]],
+    exe_args: t.Mapping[str, t.Sequence[t.Sequence[str]]],
+    _n_permutations: int = 0,
```

Comment:
I'd actually recommend revering the name change here
```suggestion
    n_permutations: int = 0,
```